// @antv/g/esm/canvas.js
import { __assign, __extends } from "tslib";
import * as Util from '@antv/util';
import { detect } from 'detect-browser';
import * as renderers from './renderers/index';
import Event from './event';
import Timeline from './core/timeline';
import Group from './core/group';
var browser = detect();
var isFirefox = browser && browser.name === 'firefox';
var EVENTS = [
    'click',
    'mousedown',
    'mouseup',
    'dblclick',
    'contextmenu',
    'mouseenter',
    'mouseleave',
    'mouseover',
    'mouseout',
    'mousemove',
    'wheel',
];
var preShape = null;
var mousedown = null;
var dragging = null;
var Canvas = /** @class */ (function (_super) {
    __extends(Canvas, _super);
    function Canvas(cfg) {
        var _this = _super.call(this, __assign({ eventEnable: true, 
            /**
             * 像素宽度
             * @type {Number}
             */
            width: null, 
            /**
             * 像素高度
             * @type {Number}
             */
            height: null, 
            /**
             * 画布宽度
             * @type {Number}
             */
            widthCanvas: null, 
            /**
             * 画布高度
             * @type {Number}
             */
            heightCanvas: null, 
            /**
             * CSS宽
             * @type {String}
             */
            widthStyle: null, 
            /**
             * CSS高
             * @type {String}
             */
            heightStyle: null, 
            /**
             * 容器DOM
             * @type {Object}
             */
            containerDOM: null, 
            /**
             * 当前Canvas的DOM
             * @type {Object}
             */
            canvasDOM: null, 
            /**
             * 屏幕像素比
             * @type {Number}
             */
            pixelRatio: null, 
            /**
             * 渲染器，默认是canvas
             * @type {String}
             */
            renderer: 'canvas', 
            /**
             * 是否支持 CSS Transform
             * CSS transform 目前尚未经过长时间验证，为了避免影响上层业务，默认关闭，上层按需开启
             */
            supportCSSTransform: false }, cfg)) || this;
        _this._setGlobalParam();
        _this._setContainer();
        _this._initPainter();
        _this._scale();
        if (_this.get('eventEnable')) {
            _this.registerEvent();
        }
        return _this;
    }
    Canvas.prototype._scale = function () {
        if (this.cfg.renderType !== 'svg') {
            var pixelRatio = this.get('pixelRatio');
            this.scale(pixelRatio, pixelRatio);
        }
    };
    Canvas.prototype._setGlobalParam = function () {
        var renderType = this.get('renderer') || 'canvas';
        if (renderType === 'svg') {
            this.set('pixelRatio', 1);
        }
        else if (!this.get('pixelRatio')) {
            this.set('pixelRatio', Util.getRatio());
        }
        this.cfg.renderType = renderType;
        var renderer = renderers[renderType];
        this.cfg.renderer = renderer;
        this.cfg.canvas = this;
        var timeline = new Timeline(this);
        this.cfg.timeline = timeline;
    };
    Canvas.prototype._setContainer = function () {
        var containerId = this.get('containerId');
        var containerDOM = this.get('containerDOM');
        if (!containerDOM && document) {
            containerDOM = document.getElementById(containerId);
            this.set('containerDOM', containerDOM);
        }
        if (containerDOM) {
            Util.modifyCSS(containerDOM, {
                position: 'relative',
            });
        }
    };
    Canvas.prototype._initPainter = function () {
        var containerDOM = this.get('containerDOM');
        var context = this.get('context');
        var painter = new this.cfg.renderer.painter(containerDOM, context);
        this.cfg.painter = painter;
        this.cfg.canvasDOM = this.cfg.el = painter.canvas;
        this.changeSize(this.get('width'), this.get('height'));
    };
    Canvas.prototype._resize = function () {
        var canvasDOM = this.get('canvasDOM');
        var widthCanvas = this.get('widthCanvas');
        var heightCanvas = this.get('heightCanvas');
        var widthStyle = this.get('widthStyle');
        var heightStyle = this.get('heightStyle');
        canvasDOM.style.width = widthStyle;
        canvasDOM.style.height = heightStyle;
        canvasDOM.setAttribute('width', widthCanvas);
        canvasDOM.setAttribute('height', heightCanvas);
    };
    Canvas.prototype.getWidth = function () {
        var pixelRatio = this.get('pixelRatio');
        var width = this.get('width');
        return width * pixelRatio;
    };
    Canvas.prototype.getHeight = function () {
        var pixelRatio = this.get('pixelRatio');
        var height = this.get('height');
        return height * pixelRatio;
    };
    Canvas.prototype.changeSize = function (width, height) {
        var pixelRatio = this.get('pixelRatio');
        var widthCanvas = width * pixelRatio;
        var heightCanvas = height * pixelRatio;
        this.set('widthCanvas', widthCanvas);
        this.set('heightCanvas', heightCanvas);
        this.set('widthStyle', width + 'px');
        this.set('heightStyle', height + 'px');
        this.set('width', width);
        this.set('height', height);
        this._resize();
    };
    // 实现接口
    Canvas.prototype.getPointByEvent = function (ev) {
        var supportCSSTransform = this.get('supportCSSTransform');
        if (supportCSSTransform) {
            var pixelRatio = this.get('pixelRatio') || 1;
            // For Firefox <= 38
            if (isFirefox && !Util.isNil(ev.layerX) && ev.layerX !== ev.offsetX) {
                return {
                    x: ev.layerX * pixelRatio,
                    y: ev.layerY * pixelRatio,
                };
            }
            if (!Util.isNil(ev.offsetX)) {
                // For IE6+, Firefox >= 39, Chrome, Safari, Opera
                return {
                    x: ev.offsetX * pixelRatio,
                    y: ev.offsetY * pixelRatio,
                };
            }
        }
        // should calculate by self for other cases, like Safari in ios
        // TODO: support CSS transform
        var _a = this.getClientByEvent(ev), clientX = _a.x, clientY = _a.y;
        return this.getPointByClient(clientX, clientY);
    };
    // 获取 touch 事件的 clientX 和 clientY 需要单独处理
    Canvas.prototype.getClientByEvent = function (ev) {
        var clientInfo = ev;
        if (ev.touches) {
            if (ev.type === 'touchend') {
                clientInfo = ev.changedTouches[0];
            }
            else {
                clientInfo = ev.touches[0];
            }
        }
        return {
            x: clientInfo.clientX,
            y: clientInfo.clientY,
        };
    };
    /**
     * 将窗口坐标转变成 canvas 坐标
     * @param  {Number} clientX 窗口x坐标
     * @param  {Number} clientY 窗口y坐标
     * @return {Object} canvas坐标
     */
    Canvas.prototype.getPointByClient = function (clientX, clientY) {
        var el = this.get('el');
        var pixelRatio = this.get('pixelRatio') || 1;
        var bbox = el.getBoundingClientRect();
        return {
            x: (clientX - bbox.left) * pixelRatio,
            y: (clientY - bbox.top) * pixelRatio,
        };
    };
    Canvas.prototype.getClientByPoint = function (x, y) {
        var el = this.get('el');
        var bbox = el.getBoundingClientRect();
        var pixelRatio = this.get('pixelRatio') || 1;
        return {
            clientX: x / pixelRatio + bbox.left,
            clientY: y / pixelRatio + bbox.top,
        };
    };
    Canvas.prototype.draw = function () {
        this.cfg.painter.draw(this);
    };
    Canvas.prototype.getShape = function (x, y, e) {
        if (arguments.length === 3 && this.cfg.renderer.getShape) {
            return this.cfg.renderer.getShape.call(this, x, y, e);
        }
        return _super.prototype.getShape.call(this, x, y);
    };
    Canvas.prototype.getRenderer = function () {
        return this.cfg.renderType;
    };
    Canvas.prototype._drawSync = function () {
        this.cfg.painter.drawSync(this);
    };
    Canvas.prototype.destroy = function () {
        var cfg = this.cfg;
        var containerDOM = cfg.containerDOM;
        var canvasDOM = cfg.canvasDOM;
        if (canvasDOM && containerDOM) {
            containerDOM.removeChild(canvasDOM);
        }
        cfg.timeline.stop();
        _super.prototype.destroy.call(this);
    };
    Canvas.prototype.registerEvent = function () {
        var self = this;
        var el = this.get('el');
        Util.each(EVENTS, function (evt) {
            el.addEventListener(evt, function (e) {
                self._triggerEvent(evt, e);
            }, false);
        });
        // special cases
        el.addEventListener('touchstart', function (e) {
            if (!Util.isEmpty(e.targetTouches)) {
                self._triggerEvent('touchstart', e);
            }
        }, false);
        el.addEventListener('touchmove', function (e) {
            if (!Util.isEmpty(e.targetTouches)) {
                self._triggerEvent('touchmove', e);
            }
        }, false);
        el.addEventListener('touchend', function (e) {
            if (!Util.isEmpty(e.changedTouches)) {
                self._triggerEvent('touchend', e);
            }
        }, false);
    };
    Canvas.prototype._getEmitter = function (element, event) {
        if (element) {
            if (Util.isEmpty(element.getEvents())) {
                var parent_1 = element.get('parent');
                if (parent_1 && !event.propagationStopped) {
                    return this._getEmitter(parent_1, event);
                }
            }
            else {
                return element;
            }
        }
    };
    Canvas.prototype._getEventObj = function (type, e, point, target) {
        var event = new Event(type, e, true, true);
        event.x = point.x;
        event.y = point.y;
        event.clientX = e.clientX;
        event.clientY = e.clientY;
        event.target = target;
        event.currentTarget = this._getEmitter(target, event);
        return event;
    };
    Canvas.prototype._triggerEvent = function (type, e) {
        var point = this.getPointByEvent(e);
        var shape = this.getShape(point.x, point.y, e);
        var el = this.get('el');
        // svg原生事件取不到dragover, dragout, drop等事件的对象。这边需要走数学拾取。
        if (dragging && this.getRenderer() === 'svg') {
            shape = this.getShape(point.x, point.y);
        }
        if (type === 'mousemove') {
            if (preShape && preShape !== shape) {
                this._emitEvent('mouseout', e, point, preShape);
                this._emitEvent('mouseleave', e, point, preShape);
                if (dragging) {
                    this._emitEvent('dragleave', e, point, preShape);
                }
                if (!preShape.destroyed && !preShape.removed) {
                    el.style.cursor = preShape.attr('cursor') || 'default';
                }
            }
            if (dragging) {
                this._emitEvent('drag', e, point, dragging);
                this._emitEvent('mousemove', e, point, shape || this);
            }
            if (shape) {
                if (!dragging) {
                    if (mousedown === shape) {
                        dragging = shape;
                        mousedown = null;
                        this._emitEvent('dragstart', e, point, shape);
                    }
                    else {
                        this._emitEvent('mousemove', e, point, shape);
                    }
                }
                if (preShape !== shape) {
                    this._emitEvent('mouseenter', e, point, shape);
                    this._emitEvent('mouseover', e, point, shape);
                    if (dragging) {
                        this._emitEvent('dragenter', e, point, shape);
                    }
                }
            }
            else {
                var canvasmousemove = this._getEventObj('mousemove', e, point, this);
                this.emit('mousemove', canvasmousemove);
            }
            preShape = shape;
        }
        else {
            this._emitEvent(type, e, point, shape || this);
            if (!dragging && type === 'mousedown') {
                mousedown = shape;
            }
            if (type === 'mouseup') {
                mousedown = null;
                if (dragging) {
                    dragging.cfg.capture = true;
                    this._emitEvent('dragend', e, point, dragging);
                    dragging = null;
                    this._emitEvent('drop', e, point, shape || this);
                }
            }
        }
        if (shape && !shape.destroyed) {
            el.style.cursor = shape.attr('cursor') || 'default';
        }
    };
    Canvas.prototype._emitEvent = function (type, evt, point, shape) {
        var event = this._getEventObj(type, evt, point, shape);
        var emitShape = this._getEmitter(shape, evt);
        emitShape && emitShape.emit(type, event);
        return emitShape;
    };
    return Canvas;
}(Group));
export default Canvas;
